### ISSUE TITLE
Query data and error are undefined

### ISSUE THREAD

**Issue by giulianok** [apollographql/apollo-client#8063]
**Intended outcome:**
I'm using `useQuery` to query and render some data. The `loading` prop works as expected, the value is `true` and then changes to `false`. When loading is false, `data` and `error` are both `undefined`.

I checked the network tab and the data is being received (I see the `data` prop)

![Screen Shot 2021-04-27 at 1 41 51 PM](https://user-images.githubusercontent.com/8062245/116288459-8c268b00-a75f-11eb-989b-95ac79b7dabb.png)

I also checked what's going on with the Apollo chrome extension and I can see the data

![Screen Shot 2021-04-27 at 1 34 24 PM](https://user-images.githubusercontent.com/8062245/116288557-a2cce200-a75f-11eb-894d-00ccc7086c1d.png)
![Screen Shot 2021-04-27 at 1 34 10 PM](https://user-images.githubusercontent.com/8062245/116288582-a8c2c300-a75f-11eb-9eda-86ccfb852327.png)

I also was able to verify the correct result from the BE using the `fetch` API

![Screen Shot 2021-04-27 at 1 34 47 PM](https://user-images.githubusercontent.com/8062245/116288813-e293c980-a75f-11eb-9619-2af73fe94dd8.png)

**Actual outcome:**
`data` and `error` props from `useQuery` are `undefined`.

**How to reproduce the issue:**

Here's the component that uses `useQuery` and `fetch`

```jsx
const QUERY = gql`
  query AccountStatus {
    accountStatus {
      __typename
      missingCode
      completed
      reason
    }
  }
`

const MissingThings = () => {
  const x = useQuery(QUERY)

  const { loading, data, error } = x

  console.log('--x', x)

  useEffect(() => {
    fetch(`${process.env.REACT_APP_GRAPHQL_URL}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization:
          'Bearer <TOKEN>',
      },
      body: JSON.stringify({
        variables: {},
        query: `
        query AccountStatus {
          accountStatus {
            __typename
            missingCode
            completed
            reason
          }
        }
        
        `,
      }),
    })
      .then((result) => result.json())
      .then((result) => console.log('--result', result))
      .catch(console.error)
  }, [])

  if (loading) {
    return null
  }

  if (error) {
    console.log('--error', error)
    return null
  }

  console.log('--data', data)

  return <div>All good</div>
}
```

And this is the Apollo Client

```jsx
const ApolloClientProvider = ({ children }: any) => {
  const { auth, account } = useGlobalProvider()
  const headers =
    auth && account ? { Authorization: `Bearer ${auth?.token}` } : {}

  console.log('--headers', headers)
  console.log('--auth', auth)

  const wsLink = new WebSocketLink({
    uri: process.env.REACT_APP_GRAPHQL_WS_URL as string,
    options: {
      reconnect: true,
      connectionParams: () => ({
        headers,
      }),
    },
  })

  const httpLink = new HttpLink({ uri: process.env.REACT_APP_GRAPHQL_URL })

  const authMiddleware = new ApolloLink((operation, forward) => {
    // add the authorization to the headers
    operation.setContext({
      headers:
        auth && account ? { Authorization: `Bearer ${auth?.token}` } : {},
    })

    return forward(operation)
  })

  const splitLink = split(
    ({ query }) => {
      const definition = getMainDefinition(query)
      return (
        definition.kind === 'OperationDefinition' &&
        definition.operation === 'subscription'
      )
    },
    wsLink,
    httpLink,
  )

  const logoutLink = onError((error) => {
    console.log('APOLLO ERROR!', error)

    if (
      error.networkError?.message.includes('JWTExpired') ||
      error.graphQLErrors?.some(
        ({ extensions, message }) =>
          extensions?.code === 'invalid-jwt' || message.includes('JWTExpired'),
      )
    ) {
      navigate('/logout')
    }
  })

  const client = new ApolloClient({
    cache: new InMemoryCache(),
    link: ApolloLink.from([logoutLink, authMiddleware, splitLink]),
  })

  return <ApolloProvider client={client}>{children}</ApolloProvider>
}
```

**Versions**
  System:
    OS: macOS 11.2.3
  Binaries:
    Node: 14.8.0 - ~/n/bin/node
    npm: 6.14.7 - ~/n/bin/npm
  Browsers:
    Chrome: 90.0.4430.72
    Safari: 14.0.3
  npmPackages:
    @apollo/client: ^3.3.15 => 3.3.15

**Comment by giulianok** [comment]
I continued testing and I found out that not using the cache works:

```js
const x = useQuery(QUERY, {
    fetchPolicy: 'network-only',
  })
```

Any idea of how to debug the cache? I see that the data is cached but Apollo is having a hard time retrieving it (that's my conclusion)

![Screen Shot 2021-04-27 at 2 03 35 PM](https://user-images.githubusercontent.com/8062245/116290414-6ef2bc00-a761-11eb-849d-94aeeafb5621.png)

**Comment by alexwasner** [comment]
I'm having the same bug of `data` being `undefined` but when backgrounding my app and foregrounding it. Verified via flipper that the data was actually coming in though. Setting `fetchPolicy: network-only` _didnt_ solve it for my case.

```
    "@apollo/client": "^3.3.15",
    "react-native": "0.63.2",
    "apollo3-cache-persist": "^0.9.1",
```

**Comment by brainkim** [comment]
I’ve also run into this situation when implementing defer for Apollo Client. It seems like if the server responds with data which doesn’t match the query, in that it’s missing fields, the result provided by the cache has `data` and `error` set to `undefined` and `loading` set to false, which should probably never happen. Here’s a quick reproduction.

```jsx
/*** SCHEMA ***/
import {
  GraphQLSchema,
  GraphQLObjectType,
  GraphQLID,
  GraphQLString,
  GraphQLList,
} from 'graphql';

const PersonType = new GraphQLObjectType({
  name: 'Person',
  fields: {
    id: {type: GraphQLID},
    name: {type: GraphQLString},
  },
});

const peopleData = [
  {id: 1, name: 'John Smith'},
  {id: 2, name: 'Sara Smith'},
  {id: 3, name: 'Budd Deey'},
];

const QueryType = new GraphQLObjectType({
  name: 'Query',
  fields: {
    people: {
      type: new GraphQLList(PersonType),
      resolve: () => peopleData,
    },
  },
});

const schema = new GraphQLSchema({ query: QueryType, mutation: MutationType });

/*** LINK ***/
import {graphql, print} from "graphql";
import {ApolloLink, Observable} from "@apollo/client";
function delay(wait) {
  return new Promise(resolve => setTimeout(resolve, wait));
}

const link = new ApolloLink(operation => {
  return new Observable(async observer => {
    const {query, operationName, variables} = operation;
    await delay(300);
    try {
      const result = await graphql(
        schema,
        print(query),
        null,
        null,
        variables,
        operationName,
      );
      // Emulate producing a result which is missing a field from the server.
      delete result.data.people[2].name;
      observer.next(result);
      observer.complete();
    } catch (err) {
      observer.error(err);
    }
  });
});

/*** APP ***/
import React from "react";
import {render} from "react-dom";
import {
  ApolloClient,
  ApolloProvider,
  InMemoryCache,
  gql,
  useQuery,
} from "@apollo/client";
import "./index.css";

const ALL_PEOPLE = gql`
  query AllPeople {
    people {
      id
      name
    }
  }
`;

function App() {
  const {loading, data, error} = useQuery(ALL_PEOPLE);
  if (loading) {
    return <div>Loading...</div>;
  } else if (error) {
    return <div>Error: {error.toString()}</div>;
  } else if (!data) {
    return <div>What the heck</div>;
  }

  return (
    <ul>
      {data.people.map(person => (
        <li key={person.id}>{person.name}</li>
      ))}
    </ul>
  );
}

const client = new ApolloClient({
  cache: new InMemoryCache(),
  link,
});

render(
  <ApolloProvider client={client}>
    <App />
  </ApolloProvider>,
  document.getElementById("root"),
);
```

**Comment by brainkim** [comment]
@benjamn Do you know the original behavior for when a server responds with less data than the query expects? My sense is that this is a regression of some kind.

**Comment by brainkim** [comment]
I’m discussing, and one thing y’all should try is adding `returnPartialData` to the query options. Still investigating as to what the expected behavior should be.

**Comment by diracs-delta** [comment]
Just noticed this bug today. Can also confirm that setting a `fetchPolicy` that does not read from cache resolves the issue. On latest version, `@apollo/client@3.3.18`.

What's really bizarre is that I've only noticed this error recently. Yet, when I downgrade to `v3.0.0`, the error persists.

> Do you know the original behavior for when a server responds with less data than the query expects? My sense is that this is a regression of some kind.

@brainkim I'm working on a new project with my company that involves a very simple query that reproduces this error consistently. I can say safely that the fields my mock server is returning is the same as the fields my client is expecting. Sadly, the code is private so I can't give you the repo.

Basically I'm reproducing this error when I run 2 different queries that have the same cache key. I have a 

```
useQuery(GET_USERS)
```

called earlier, which returns an array of all users including one with ID `asdf`, and a 

```
useQuery(SELECT_USER, { variables: { id: 'asdf' } })
```

called later. The second `useQuery` fails silently and eventually resolves to `{ loading: false, error: false, data: undefined }`.

**Comment by diracs-delta** [comment]
I was able to reproduce this issue on a separate repository. See below @brainkim 

https://github.com/diracs-delta/apollo-client-8063

**Comment by brainkim** [comment]
@diracs-delta This is a beautiful example. I’m dropping everything to investigate.

**Comment by brainkim** [comment]
@diracs-delta Ah I get it. It’s the same problem as I think everyone having this issue is having. In the mock graphql server you need to put the `user object under the key `user`, or it doesn’t actually match the GraphQL query you sent to the server.

```js
const handlers = [
  graphql.query('getUsers', (_, res, ctx) => {
    return res(ctx.data({ users: userStore }))
  }),
  graphql.query('getSpecificUser', (req, res, ctx) => {
    const user = userStore.find(user => user.id === req.variables.id);

    if (!user) {
      return res(ctx.errors([{ message: `Cannot find user ${user}` }]));
    }

    return res(ctx.data({user}));
  }),
];
```

**Comment by diracs-delta** [comment]
@brainkim I believe you're right. The `useQuery` hook is silently failing when the response doesn't match the query exactly, but the response has a corresponding cache entry.

It would be nice to produce some kind of console warning or emit an error when this happens to prevent future confusion. I can take a further look at this later tonight.

**Comment by jasongaare** [comment]
Hello, I'm experiencing this same issue, however I am _not_ mocking anything. 

As @brainkim suggested above, adding `returnPartialData` allows my `useQuery` hook to work again and properly return data. However, I'm not keen on adding that to all my queries with caching involved? 

Is there somewhere in the cache setup where this can be addressed? Or what long term solution could I implement that doesn't involve adding the `returnPartialData` option to all my useQuery hooks?

**Comment by diracs-delta** [comment]
@jasongaare 

If you are both the provider and consumer of the GraphQL API, this issue is occurring because the data structure of the query and the response do not match, which is usually due to the backend omitting a null field entirely instead of returning `null` for a null field.

Otherwise, if you are only the consumer of the GraphQL API, you're unfortunately out of other options until this issue is resolved.

**Comment by macrael** [comment]
Where is the requirement to return "null" instead of "undefined" documented? Is that a graphql spec kind of thing? I hit this hard when I tried to change my Maybe type to `T | undefined` instead of `T | null` in graphql code-gen. Is that just not going to work on the client b/c the spec requires "null"?

**Comment by diracs-delta** [comment]
@macrael Although I truthfully haven't read the spec top-to-bottom, I believe that to be the case. I think the provider of a GraphQL API should always explicitly return `null` for a field rather than omitting it in the response, but someone please correct me if I'm wrong.

The reason codegen defines `Maybe<T> = T | undefined | null` by default is because codegen is intended to be used primarily for **client-side** development, in which case the type of `Maybe` is very convenient. This way, in queries with optional parameters, you can just omit the optional parameters instead of explicitly specifying each to be `null`. This omission is definitely in the GraphQL spec for **request variables, but not for response data**.

**Comment by DeannaWang** [comment]
I've also encountered this issue, and setting `fetchPolicy: network-only` is not fixing it. The oddest thing is that when I'm on a page where an object is queried for some fields, and then I open a component where the same object is queried for different fields, it is **WORKING**. However, when I'm on a page that doesn't contain any graphql query, and I open the same component that queries the object, the issue happens. In the network tab in dev tools the API actually returned the right data, but in the code where I did `await client.query` I got undefined. I've tried to add `ID` to each of the object and subobjects, passing in `fetchPolicy: no-cache` or `fetchPolicy: network-only`, and nothing works.

**Comment by hwillson** [comment]
This should be resolved in @apollo/client@latest - let us know if not.

**Comment by alimoli** [comment]
Trying to install the package `@apollo/client@latest`.

Yarn version: 
 `1.22.10` 

Node version: 
  `14.17.4`

Platform: 
  `darwin x64`

```
Trace: 
  Invariant Violation: expected manifest
      at invariant (/usr/local/Cellar/yarn/1.22.10/libexec/lib/cli.js:2314:15)
      at PackageResolver.getStrictResolvedPattern (/usr/local/Cellar/yarn/1.22.10/libexec/lib/cli.js:65361:5)
      at add (/usr/local/Cellar/yarn/1.22.10/libexec/lib/cli.js:94632:33)
      at add (/usr/local/Cellar/yarn/1.22.10/libexec/lib/cli.js:94673:9)
      at add (/usr/local/Cellar/yarn/1.22.10/libexec/lib/cli.js:94673:9)
      at PackageHoister.prepass (/usr/local/Cellar/yarn/1.22.10/libexec/lib/cli.js:94706:7)
      at PackageHoister.seed (/usr/local/Cellar/yarn/1.22.10/libexec/lib/cli.js:94098:10)
      at PackageLinker.getFlatHoistedTree (/usr/local/Cellar/yarn/1.22.10/libexec/lib/cli.js:48740:13)
      at PackageLinker.<anonymous> (/usr/local/Cellar/yarn/1.22.10/libexec/lib/cli.js:48755:27)
      at Generator.next (<anonymous>)
```

**Comment by stefandplac** [comment]
I have the solution just to get the data out . My solution is this one. (it is temporary) :)) I wait for the bug to be solved

const {loading, error, data} = useQuery(queries.GET_CATEGORIES); //at this moment the data is undefined
 //using the spread operator I create a new object with the data content
  const prod = {...data};
  console.log(prod);

**Comment by ucarbehlul** [comment]
I can confirm that this is still happening at 3.5.10. If one of the fields requested is null (because of an error), both errors and data fields are undefined from useQuery returned object.

**Comment by mvincent7891** [comment]
I'm seeing the same thing (@apollo/client 3.5. 10):

 `useQuery` returns `undefined` for `error` and `data`, though I can see that `loading` is true at some point.
 
We both consume and provide the GraphQL API, so I full ability to debug on both ends. The network response is exactly as we'd expect. Notable, there are no errors, and the shape of the response is exactly as we'd expected, with the object nested under `data`.

For now, we've downgraded back to 3.2.5 to resolve this.

**Comment by martinmolina1988** [comment]
I had the same problem, the base url was wrong
export const client = new ApolloClient({
  uri: process.env.NEXT_PUBLIC_BASE_URL + "/api/graphql",
  cache,
});

**Comment by Bruce773** [comment]
I just ran into this problem, as well. Fixed it by downgrading to `3.2.5`.

**Comment by fvctr** [comment]
I just ran into this issue too...

**Comment by akbarnurullaev** [comment]
Same problem 🥲

**Comment by jcgonzave** [comment]
Same issue happening at 3.6.1.

**Comment by minnocci-bcgdv** [comment]
3.5.10 works for me 🔥 

In addition, my snapshot tests using `"@apollo/client": "3.5.10"` and  `"@testing-library/react": "^13.1.1"` look like:

`
it('Renders correctly for a successful call', async () => {
    const { container } = render(
      <MockedProvider
        defaultOptions={{ watchQuery: { fetchPolicy: 'network-only' } }}
        mocks={[mocksNeeded]}
        addTypename={false}
      >
        <AllTheChildren/>
      </MockedProvider>,
    )
    await waitFor(() => new Promise((res) => setTimeout(res, 500)))
    expect(container).toMatchSnapshot()
  })
`

**Comment by 87brandonn** [comment]
same problem here. fixed by downgrading to 3.2.5

**Comment by alex-nguyen90** [comment]
downgrading to 3.2.5 works for me

**Comment by w0o** [comment]
Tried a few suggestions here, I can confirm that both 3.2.5 and 3.5.10 are working for me and stayed with 3.5.10.

I'm really curious also as to why no errors are being reported and we end up with a silent fail, this has been very time-consuming to debug.

Maybe this suggests a regression, let us know if there is anything else we can do to test or provide further debug info.

**Comment by bentron2000** [comment]
3.5.10 working for me. No joy with 3.6+

**Comment by minnocci-bcgdv** [comment]
3.6.0 also failing for me, sticking to 3.5.10

**Comment by jeffreyyoung** [comment]
3.3.7 was failing for me.  Fixed by upgrading to 3.5.10.
```diff
-    "@apollo/client": "3.3.7",
+    "@apollo/client": "3.5.10",
```

**Comment by eranimo** [comment]
Why is this closed?

**Comment by Maxth** [comment]
+1

**Comment by Pastbin** [comment]
it seems I was able to solve the problem!

If your server responds with a status other than 200, then "data" becomes "undefined"
Example:
        response.status(403)
return {message: "not authorized"}
data (client) = undefined

        response.status(200)
(default 200)
return {message: "not authorized"}
data (client) = data{message: "not authorized"}

I hope this helps someone, good luck!

**Comment by anonymous-coder** [comment]
I'm a little shocked this is still a thing, it only worked on 3.3.7 for me, not even 3.5.10.

**Comment by prayutsu** [comment]
This isn't resolved yet. I tried downgrading to 3.2.5, 3.5.10 but didn't work for me.

**Comment by Saad9624** [comment]
getting undefined for data and error

**Comment by MorganeLecurieux** [comment]
Same issue for me on `3.6.9`

**Comment by jpwexperience** [comment]
Running into this problem on `3.7.1`

First query gives undefined for data and error. Response of 200 with expected data can be seen in the network.

**Comment by jimmygizmo** [comment]
I've tried every suggestion in this post and many other sources and most of the apollo client versions that people said worked and the problem is present 100% of the time. My schema and query are as simple as possible. The shape matches perfectly. ** The correct data is clearly being returned during the fetch. I have lost a week over this and have no choice but to give up on Apollo. I don't agree that anyone in this thread has identified the root cause or actual working versions. It seems to me the root cause has not been touched on yet in this thread. This bug is a deal breaker. Apollo is getting ripped out.

**Comment by hwillson** [comment]
@jimmygizmo can you provide a small runnable reproduction that shows this happening?

**Comment by RobJJ** [comment]
Still getting this issue: v3.7.9

**Comment by Sabplpz** [comment]
Happening at 3.7.17, tried downgrading to 3.5.10 to test and because I'm using React 18.2 it wouldn't let me downgrade. Not sure what to do, any suggestions?

**Comment by Eric-Fray** [comment]
Confirmed the same on version 3.80 and 3.8.1. Downgrading is also not an option for me.

**Comment by Rishab-Rao-G** [comment]
I know I'm super late to this. Was facing the same issue recently. 
But adding `errorPolicy: 'all'` worked for me. 


```  
 const { data, error, loading } = useQuery(query, {
    errorPolicy: 'all',
    client,
    variables,
    skip,
  });
```

Before - 
<img width="744" alt="Screenshot 2024-08-29 at 18 31 03" src="https://github.com/user-attachments/assets/b50fa4be-302c-4687-be0e-acb7d62ac67d">


After adding errorPolicy - 
<img width="608" alt="Screenshot 2024-08-29 at 18 31 28" src="https://github.com/user-attachments/assets/f839d348-5421-4a56-b937-a99672708bd0">

**Comment by Joshrogan** [comment]
Upgrading to 3.9.11 fixed this for me.

**Comment by WildDanDan** [comment]
For anyone using Hasura, I solved this issue (in `3.11.8`) by setting the environment variable `HASURA_GRAPHQL_EXPERIMENTAL_FEATURES=no_null_unbound_variable_default`. Look for similar settings in your GraphQL engine. Hope this helps someone.
